### DISCUSSION TITLE
Federated validation without training

### DISCUSSION THREAD

**Discussion by hitesh-ag1** [NVIDIA/NVFlare#1490] [Q&A] [answered]
Hi, I am trying to understand how I can just run federated validation by uploading model weights without undergoing through the training pipelines using NVFlare. Any advice or help would be appreciated!

**Comment by chesterxgchen** [comment, marked as answer]
There are two issues to do this: 

1) This requires we able to identify a particular workflow ( cross-validation workflow) and only execute that workflow. 
Currently, we have to change the code to do that. We actually have a PR: https://github.com/NVIDIA/NVFlare/pull/1542

Once you have above code in the PR, you should be able add "target_workflow_id":"cross_site_model_eval", then run the cross validation. 

2) Second issues is to make sure the next run don't overwrite the model. 
For example, the current simulator run using a fixed job id = "simulator_job", before each simulator run, the content under job_id is removed.  If the model persistence path is under Job_id, then it will work. But for normal job (POC or production), they don't have this issue. You can upload the model and re-run cross validation

This PR will likely merged after 2.3.0 release and get release in Summar release. If you can't wait, you can use above PR to patch your local code.

**Reply by hitesh-ag1** [reply]
Alright, thanks for the guidance! I'll try the above method.

**Comment by chesterxgchen** [comment]
Just noticed I had a typo, let me iterate

for simulator,  the content under job id, (job id = simulator_job) is wipe out before each run. This includes the model. If the model still save under the job id , running validation without training wont work, because the model will be deleted before job starts. 

So, you need to either make sure the model wont save under job id or simulator wont delete the model.

There several ways to do it:
 1) change the simulator so that for each run it change the job Id to different directory such simulator_job01, 02, etc

2) don’t change simulator, but change model persistor to save a model to different location and change the model loader to load from this new location 

3) don’t use simulator, use poc mode or non-HA , local production mode. No need to change code

**Reply by hitesh-ag1** [reply]
Alright, finally gave this a try today. Thanks for the guidance @chesterxgchen. I used the PR #1542 in local production mode and modified the config_fed_server.json as described. After this, I had to make some changes to the _cross_site_model_eval_ workflow along with the _pt_file_model_locator_ to use a user specified model weights file from a specified directory for the validation process. Following this, everything worked successfully.

**Reply by chesterxgchen** [reply]
Great to hear that. we will try to get the PR merged in the next release. so you don't need to customize anymore. stay tuned.

**Reply by chesterxgchen** [reply]
follow-up question for you: 
This is one way to solve the problem, did you considered other alternative approaches, such as
1) define a new job with validation workflow ?  ( copy the existing job and remove training workflow ) 
2) simply remove the training workflow in the current job config and re-run job  ?

**Reply by hitesh-ag1** [reply]
So my usecase is that an external user (**not one of the sites**) uploads his/her model weights on the master server for validation across all sites in the federated network. 

I tried the mentioned methods for my usecase, however there were two issues with the current validation workflow (i.e. _cross_site_model_eval_):
1. It has a cleanup function defined within the file where it cleans up the model directory and creates a new one before starting validation. This essentially removes the uploaded model by external user on the master server for validation.
2. After removing the model file from master server, it communicates with the sites and obtains their local models for validation (which is not my usecase).

If my usecase had been validation of models existing at the sites, then the alternative approaches you mentioned would have worked directly without any further changes.

**Reply by YuanTingHsieh** [reply]
@hitesh-ag1 thanks for your feedback!

We will consider this use case and hope we can add an example workflow in the next release.

**Reply by YuanTingHsieh** [reply]
@hitesh-ag1 with example in here: https://github.com/NVIDIA/NVFlare/pull/2035
You will be able to run cross-site validation WITHOUT training.

I know this does not match 100% your use case just yet.
We will be developing examples related to your use case "an external user (not one of the sites) uploads his/her model weights on the master server for validation across all sites in the federated network."

You can also write your own workflows at the same time.
The logic should be quite easy:
1. Initialize/Load model from server side (the external user upload his/her model to server side already)
2. Send this model to all client sites to evaluate
3. Get back evaluation results
